## Homework - 3
#### Submitted by Apurv Mittal

In [1]:
import numpy as np
from scipy.stats import rankdata
from scipy import stats

                            

### You asked your 10 work friends to answer a survey. They gave you back the following dictionary object.  

In [2]:

people = {'Jane': {'willingness to travel':1,
                  'desire for new experience':5,
                  'cost':4,
                  'food':2,
                  'hipster points':7,
                  'vegetarian':8 },
          'Liz': {'willingness to travel':8,
                  'desire for new experience':4,
                  'cost':6,
                  'food':5,
                  'hipster points':2,
                  'vegetarian':6 },
          'Mark': {'willingness to travel':4,
                  'desire for new experience':6,
                  'cost':1,
                  'food':3,
                  'hipster points':6,
                  'vegetarian':7 },
          'Sharon':{'willingness to travel':7,
                  'desire for new experience':4,
                  'cost':1,
                  'food':4,
                  'hipster points':6,
                  'vegetarian':3 },
          'Sergio':{'willingness to travel':2,
                  'desire for new experience':3,
                  'cost':5,
                  'food':4,
                  'hipster points':6,
                  'vegetarian':1 },
          'Sarah':{'willingness to travel':9,
                  'desire for new experience':8,
                  'cost':7,
                  'food':6,
                  'hipster points':4,
                  'vegetarian':3 },
          'Ron': {'willingness to travel':7,
                  'desire for new experience':4,
                  'cost':1,
                  'food':5,
                  'hipster points':2,
                  'vegetarian':9 },
          'Apurv': {'willingness to travel':3,
                  'desire for new experience':6,
                  'cost':9,
                  'food':5,
                  'hipster points':8,
                  'vegetarian':7 },
          'Cynthia':{'willingness to travel':1,
                  'desire for new experience':4,
                  'cost':7,
                  'food':5,
                  'hipster points':2,
                  'vegetarian':9 },
          'Camden':{'willingness to travel':3,
                  'desire for new experience':2,
                  'cost':1,
                  'food':4,
                  'hipster points':6,
                  'vegetarian':9 }
                    
                }          


#### In above dictionary, replaced 'Indian Food' and 'Mexican Food' with just 'Food', since the restaurant data has only 6 features.

In [3]:
people

{'Jane': {'willingness to travel': 1,
  'desire for new experience': 5,
  'cost': 4,
  'food': 2,
  'hipster points': 7,
  'vegetarian': 8},
 'Liz': {'willingness to travel': 8,
  'desire for new experience': 4,
  'cost': 6,
  'food': 5,
  'hipster points': 2,
  'vegetarian': 6},
 'Mark': {'willingness to travel': 4,
  'desire for new experience': 6,
  'cost': 1,
  'food': 3,
  'hipster points': 6,
  'vegetarian': 7},
 'Sharon': {'willingness to travel': 7,
  'desire for new experience': 4,
  'cost': 1,
  'food': 4,
  'hipster points': 6,
  'vegetarian': 3},
 'Sergio': {'willingness to travel': 2,
  'desire for new experience': 3,
  'cost': 5,
  'food': 4,
  'hipster points': 6,
  'vegetarian': 1},
 'Sarah': {'willingness to travel': 9,
  'desire for new experience': 8,
  'cost': 7,
  'food': 6,
  'hipster points': 4,
  'vegetarian': 3},
 'Ron': {'willingness to travel': 7,
  'desire for new experience': 4,
  'cost': 1,
  'food': 5,
  'hipster points': 2,
  'vegetarian': 9},
 'Apurv': 

### Replace the keys in 'People' dictionary to match Restauant 'Features' 

'willingness to travel' ='distance',
'desire for new experience'='novelty',
'cost'='cost',
'food'='cuisine',
'hipster points'='average rating',
'vegetarian'='vegetarians'

In [4]:
people_new = {'Jane': {'distance':1,
                  'novelty':5,
                  'cost':4,
                  'cuisine':2,
                  'average rating':7,
                  'vegetarians':8 },
          'Liz': {'distance':8,
                  'novelty':4,
                  'cost':6,
                  'cuisine':5,
                  'average rating':2,
                  'vegetarians':6 },
          'Mark': {'distance':4,
                  'novelty':6,
                  'cost':1,
                  'cuisine':3,
                  'average rating':6,
                  'vegetarians':7 },
          'Sharon':{'distance':7,
                  'novelty':4,
                  'cost':1,
                  'cuisine':4,
                  'average rating':6,
                  'vegetarians':3 },
          'Sergio':{'distance':2,
                  'novelty':3,
                  'cost':5,
                  'cuisine':4,
                  'average rating':6,
                  'vegetarians':1 },
          'Sarah':{'distance':9,
                  'novelty':8,
                  'cost':7,
                  'cuisine':6,
                  'average rating':4,
                  'vegetarians':3 },
          'Ron': {'distance':7,
                  'novelty':4,
                  'cost':1,
                  'cuisine':5,
                  'average rating':2,
                  'vegetarians':9 },
          'Apurv': {'distance':3,
                  'novelty':6,
                  'cost':9,
                  'cuisine':5,
                  'average rating':8,
                  'vegetarians':7 },
          'Cynthia':{'distance':1,
                  'novelty':4,
                  'cost':7,
                  'cuisine':5,
                  'average rating':2,
                  'vegetarians':9 },
          'Camden':{'distance':3,
                  'novelty':2,
                  'cost':1,
                  'cuisine':4,
                  'average rating':6,
                  'vegetarians':9 }
                    
                }          


In [5]:
restaurants  = {'Flacos':{'distance':9,
                        'novelty':8,
                        'cost':7,
                        'cuisine':5,
                        'average rating':6, 
                        'vegetarians':4},
                'Windmills':{'distance':9,
                        'novelty':6,
                        'cost':5,
                        'average rating':9, 
                        'cuisine':8,
                        'vegetarians':5},
                'Abuelos':{'distance':7,
                        'novelty':1,
                        'cost':9, 
                        'cuisine':4,
                        'average rating':8,
                        'vegetarians':2},
                'BestThai':{'distance':9,
                        'novelty':7,
                        'cost':9,
                         
                        'cuisine':7,
                        'average rating':8,
                        'vegetarians':5},
                'SpiceRack':{'distance':9,
                        'novelty':1,
                        'cost':9,
                         
                        'cuisine':1,
                        'average rating':2,
                        'vegetarians':3},
                'PFChang':{'distance':1,
                        'novelty':7,
                        'cost':5,
                        
                        'cuisine':7,
                        'average rating':2, 
                        'vegetarians':3},
                'Hard8':{'distance':4,
                        'novelty':6,
                        'cost':4,
                        
                        'cuisine':1,
                        'average rating':8, 
                        'vegetarians':1},
                'Chilis':{'distance':2,
                        'novelty':5,
                        'cost':6,
                        
                        'cuisine':1,
                        'average rating':5, 
                        'vegetarians':3},
                'NorthItalia':{'distance':2,
                        'novelty':8,
                        'cost':3,
                         
                        'cuisine':1,
                        'average rating':4,
                        'vegetarians':2},
                'Kai':{'distance':6,
                        'novelty':9,
                        'cost':2,
                        
                        'cuisine':7,
                        'average rating':9, 
                        'vegetarians':1},
               }


In [6]:
people_names = list(people_new.keys())
people_names

['Jane',
 'Liz',
 'Mark',
 'Sharon',
 'Sergio',
 'Sarah',
 'Ron',
 'Apurv',
 'Cynthia',
 'Camden']

In [7]:
rests_names = list(restaurants.keys())
rests_names

['Flacos',
 'Windmills',
 'Abuelos',
 'BestThai',
 'SpiceRack',
 'PFChang',
 'Hard8',
 'Chilis',
 'NorthItalia',
 'Kai']

In [8]:
features_names = list(set([y for x in restaurants.keys() for y in restaurants[x].keys()]))
features_names

['cost', 'cuisine', 'novelty', 'average rating', 'distance', 'vegetarians']

In [9]:
X = np.zeros((len(features_names),len(people_names)))



In [10]:
# Reference: Code shared in the live session by prof. Nedelina Teneva

# Transform the user data into a matrix(M_people). Keep track of column and row ids.
for x in people_names:
  for f in features_names:
    if f in people_new[x]:
      X[features_names.index(f),people_names.index(x)] = people_new[x][f] 

In [11]:
print(X)

[[4. 6. 1. 1. 5. 7. 1. 9. 7. 1.]
 [2. 5. 3. 4. 4. 6. 5. 5. 5. 4.]
 [5. 4. 6. 4. 3. 8. 4. 6. 4. 2.]
 [7. 2. 6. 6. 6. 4. 2. 8. 2. 6.]
 [1. 8. 4. 7. 2. 9. 7. 3. 1. 3.]
 [8. 6. 7. 3. 1. 3. 9. 7. 9. 9.]]


In [12]:
# Transform the restaurant data into a matrix(M_resturants) use the same column index.
A = np.zeros((len(rests_names), len(features_names)))
for x in rests_names:
  for f in features_names:
    if f in restaurants[x]:
      A[rests_names.index(x),features_names.index(f)] = restaurants[x][f]

In [13]:
print(A)

[[7. 5. 8. 6. 9. 4.]
 [5. 8. 6. 9. 9. 5.]
 [9. 4. 1. 8. 7. 2.]
 [9. 7. 7. 8. 9. 5.]
 [9. 1. 1. 2. 9. 3.]
 [5. 7. 7. 2. 1. 3.]
 [4. 1. 6. 8. 4. 1.]
 [6. 1. 5. 5. 2. 3.]
 [3. 1. 8. 4. 2. 2.]
 [2. 7. 9. 9. 6. 1.]]


#### Informally describe what a linear combination is  and how it will relate to our restaurant matrix.

Linear Combination is an expression in which each term (from a set of terms)is multiplyed by a constant and then all the values are adding to get a final value. It can be represented as `AX=B` where X is a matrix, A is a set of contants when multiplyed and added we get `B`. 

In restaurant matrix, we are multiplying the restaurant scores for set of features to that of the people matrix with score for each feature. The total score `B` is obtained for each individual. 
                  
                  
Reference: https://en.wikipedia.org/wiki/Linear_combination

In [14]:
# Calculate B by taking a dot product of Restaurants (A) and People (X)
B = np.dot(A,X)

#### Choose a person and compute(using a linear combination) the top restaurant for them.  What does each entry in the resulting vector represent? 

Selecting the user name as `Camden`

Score for each restaurant

In [15]:
# Selecting user named 'Camden'

for i in people_names:
    if i == 'Camden':
        v=B[people_names.index('Camden')]
        print (v)

[144. 155. 162. 165. 132. 221. 142. 204. 118. 129.]


The above matrix is the total score for `Camden` for each of the restaurants

In [16]:
# Maximum score
np.max(v)

221.0

Index of the restaurant for which `Camden` has highest score

In [17]:
idx= np.where(v == v.max())

In [18]:
idx

(array([5]),)

##### Name of the restaurant for which Camden has highest score

In [19]:
rests_names[5]

'PFChang'


### Next, compute a new matrix (M_usr_x_rest  i.e. an user by restaurant) from all people.  What does the a_ij matrix represent? 


In [20]:
# Calculate B by taking a dot product of Restaurants (A) and People (X)
B = np.dot(A,X)
B

array([[161., 207., 170., 170., 137., 260., 175., 239., 163., 142.],
       [178., 214., 190., 193., 152., 263., 195., 255., 171., 175.],
       [128., 162., 117., 132., 128., 196., 116., 206., 128., 114.],
       [190., 235., 191., 191., 165., 289., 196., 284., 196., 171.],
       [ 90., 157.,  87., 101.,  85., 175., 112., 156., 112.,  81.],
       [108., 123., 105.,  89.,  91., 159., 106., 162., 130.,  89.],
       [116., 107., 114., 111.,  99., 153.,  86., 160.,  86.,  89.],
       [112., 105.,  98.,  83.,  86., 135.,  82., 156., 106.,  83.],
       [100.,  91., 100.,  83.,  73., 131.,  80., 132.,  86.,  71.],
       [144., 155., 162., 165., 132., 221., 142., 204., 118., 129.]])

The above matrix represents the total score of each user against each restaurant based on the various features.

### Sum all columns in M_usr_x_rest to get the optimal restaurant for all users.  What do the entries represent?

Get the sum of each column (representing restaurants) to find the restaurants with highest score.

In [21]:
rest_score = np.sum(B, axis=0)
rest_score

array([1327., 1556., 1334., 1318., 1148., 1982., 1290., 1954., 1296.,
       1144.])

In [22]:
# Maximum score
np.max(rest_score)

1982.0

The maximum score of a restaurant is `1982`. This restaurant appears to be most preferred by everyone when considering all the features.

In [23]:
# Index of the restaurant
idx= np.where(rest_score == rest_score.max())
idx

(array([5]),)

In [24]:
idx

(array([5]),)

##### Name of the restaurant for which has highest overall score

In [25]:
rests_names[5]

'PFChang'

In [26]:
top_rest_loc = np.where(np.max(B.T)==B.T)[0][0]

In [27]:
rests_names[top_rest_loc]

'PFChang'

### Now convert each row in the M_usr_x_rest into a ranking for each user and call it M_usr_x_rest_rank.   Do the same as above to generate the optimal restaurant choice.  

In [28]:
def get_real_rank(data):
    return rankdata(len(data) - rankdata(data))

In [29]:
M_usr_x_rest_rank = rankdata((B.T))
M_usr_x_rest_rank

array([ 65. ,  79. ,  44. ,  80.5,  18. ,  31. ,  38.5,  34. ,  23.5,
        56. ,  91. ,  92. ,  67. ,  94. ,  62. ,  42. ,  30. ,  26.5,
        19.5,  59. ,  72.5,  80.5,  40. ,  82.5,  14. ,  26.5,  36.5,
        21. ,  23.5,  67. ,  72.5,  84. ,  50. ,  82.5,  25. ,  16. ,
        32. ,   7. ,   7. ,  70.5,  53. ,  57. ,  44. ,  70.5,   9. ,
        19.5,  22. ,  11.5,   2. ,  50. ,  97. ,  98. ,  87. , 100. ,
        77. ,  63. ,  58. ,  52. ,  48. ,  93. ,  77. ,  85. ,  38.5,
        87. ,  34. ,  28.5,  11.5,   5. ,   3. ,  54.5,  95. ,  96. ,
        90. ,  99. ,  60.5,  67. ,  64. ,  60.5,  50. ,  89. ,  69. ,
        74.5,  44. ,  87. ,  34. ,  47. ,  11.5,  28.5,  11.5,  41. ,
        54.5,  77. ,  36.5,  74.5,   4. ,  16. ,  16. ,   7. ,   1. ,
        46. ])

### Why is there a difference between the two?  What problem arrives?  What does it represent in the real world?

In terms of rank the lowest number is better but for the rest of the data, higher number is considered better.

Some users may skew the results based on their score for particular feature.

### Find  user profiles that are problematic, explain why?

In [30]:
print("Sum:\n",np.sum(B, axis=1))

Sum:
 [1824. 1986. 1427. 2108. 1156. 1162. 1121. 1046.  947. 1572.]


Some users have much higher total scores for the restaurants for example user at Index `0` has total of `1824` while user at `8` has only `947`. This may skew the results of ranking based on the preferences of a particular user.

### How should you preprocess your data to remove this problem. 

We can normalize the data by dividing the each score by the sum of all the scores of each user.

In [31]:
row_sums = B.sum(axis=1)
normalized_B= B / row_sums[:, np.newaxis]
#reference: https://stackoverflow.com/questions/8904694/how-to-normalize-a-2-dimensional-numpy-array-in-python-less-verbose

In [32]:
normalized_B

array([[0.08826754, 0.11348684, 0.09320175, 0.09320175, 0.07510965,
        0.14254386, 0.09594298, 0.1310307 , 0.08936404, 0.07785088],
       [0.08962739, 0.10775428, 0.09566969, 0.09718026, 0.07653575,
        0.13242699, 0.09818731, 0.12839879, 0.08610272, 0.08811682],
       [0.08969867, 0.11352488, 0.08199019, 0.09250175, 0.08969867,
        0.13735109, 0.08128942, 0.14435879, 0.08969867, 0.07988788],
       [0.09013283, 0.11148008, 0.09060721, 0.09060721, 0.07827324,
        0.13709677, 0.09297913, 0.13472486, 0.09297913, 0.08111954],
       [0.07785467, 0.13581315, 0.07525952, 0.08737024, 0.07352941,
        0.15138408, 0.09688581, 0.1349481 , 0.09688581, 0.0700692 ],
       [0.0929432 , 0.10585198, 0.09036145, 0.07659208, 0.07831325,
        0.13683305, 0.09122203, 0.1394148 , 0.11187608, 0.07659208],
       [0.10347904, 0.09545049, 0.10169492, 0.09901873, 0.08831401,
        0.13648528, 0.07671722, 0.14272971, 0.07671722, 0.0793934 ],
       [0.10707457, 0.10038241, 0.0936902

### Think of two metrics to compute the disatistifaction with the group.  



`1` Find the top most ranked feature by the user and compare the score of that against that of the restaurant. If its positive then user is satisfied but if its in negative which means the user gave it higher weight but the restaurant score was lower then that can be used as the dissatisfaction score.

`2` Calculate the restaurant average score and subtract the each users restaurant score. The difference between these values could be considered a “dissatisfaction score” if the value is positive (average restaurant score is higher than the users score) for each person.

### Should you split in two groups today? 

We can split the group into two, however once we normalize the data that is not required.

If we have to split then we can take a threshold of the total score of a user like all users under 1300 total score in once group and rest in another.

### Ok. Now you just found out the boss is paying for the meal. How should you adjust? Now what is the best restaurant?

If boss is paying that means the cost criteria for all users should be minimal to `0`

In [33]:
people_cost0 = {'Jane': {'distance':1,
                  'novelty':5,
                  'cost':0,
                  'cuisine':2,
                  'average rating':7,
                  'vegetarians':8 },
          'Liz': {'distance':8,
                  'novelty':4,
                  'cost':0,
                  'cuisine':5,
                  'average rating':2,
                  'vegetarians':6 },
          'Mark': {'distance':4,
                  'novelty':6,
                  'cost':0,
                  'cuisine':3,
                  'average rating':6,
                  'vegetarians':7 },
          'Sharon':{'distance':7,
                  'novelty':4,
                  'cost':0,
                  'cuisine':4,
                  'average rating':6,
                  'vegetarians':3 },
          'Sergio':{'distance':2,
                  'novelty':3,
                  'cost':0,
                  'cuisine':4,
                  'average rating':6,
                  'vegetarians':1 },
          'Sarah':{'distance':9,
                  'novelty':8,
                  'cost':0,
                  'cuisine':6,
                  'average rating':4,
                  'vegetarians':3 },
          'Ron': {'distance':7,
                  'novelty':4,
                  'cost':0,
                  'cuisine':5,
                  'average rating':2,
                  'vegetarians':9 },
          'Apurv': {'distance':3,
                  'novelty':6,
                  'cost':0,
                  'cuisine':5,
                  'average rating':8,
                  'vegetarians':7 },
          'Cynthia':{'distance':1,
                  'novelty':4,
                  'cost':0,
                  'cuisine':5,
                  'average rating':2,
                  'vegetarians':9 },
          'Camden':{'distance':3,
                  'novelty':2,
                  'cost':0,
                  'cuisine':4,
                  'average rating':6,
                  'vegetarians':9 }
                    
                }    

In [34]:
# Reference: Code shared in the live session by prof. Nedelina Teneva

X_new = np.zeros((len(features_names),len(people_names)))

# Transform the new user data into a matrix with cost as 0.
for x in people_names:
  for f in features_names:
    if f in people_cost0[x]:
      X_new[features_names.index(f),people_names.index(x)] = people_cost0[x][f] 

In [35]:
X_new

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [2., 5., 3., 4., 4., 6., 5., 5., 5., 4.],
       [5., 4., 6., 4., 3., 8., 4., 6., 4., 2.],
       [7., 2., 6., 6., 6., 4., 2., 8., 2., 6.],
       [1., 8., 4., 7., 2., 9., 7., 3., 1., 3.],
       [8., 6., 7., 3., 1., 3., 9., 7., 9., 9.]])

In [36]:
# Calculate B_new by taking a dot product of Restaurants (A) and People (X_new)
B_new = np.dot(A,X_new)

In [37]:
top_rest_loc_new = np.where(np.max(B_new)==B_new)[0][0]

In [38]:
rests_names[top_rest_loc_new]

'Windmills'

Top restaurant name is Windmills after the cost factor is reduced.

### Tomorrow you visit another team. You have the same restaurants and they told you their optimal ordering for restaurants.  Can you find their weight matrix? 

By taking a dot product of restaurant matrix with ordering matrix transposed we can find the weight matrix.